### PR TITLE
Self verifying executable tests

### PR DESCRIPTION
**Pull request by taylus**
WIP PR for #184

### CODE REVIEW

**Comment by GrahamTheCoder** [comment]
I've added a change needed to compile on my machine.
I've also added the characterization that passes on my machine. Could you check whether it works for you locally? It's a slightly odd issue where the System reference can't be resolved, so if it's working differently on your machine we'll need either more investigation.

**Comment by taylus** [comment]
:D

Those changes actually cause the VB -> C# solution and project tests to fail on my end. The "Attribute" name doesn't appear in the attribute when it converts. Weird -- I wonder what the build server will do.

image

File: Tests/CSharp/SelfVerifyingTests.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            } catch (CompilationException ex) {
+                throw new XunitException($"Error compiling target: {ex}");
+            }
+        }
+
+        private static IEnumerable<MetadataReference> AdditionalReferences => new List<MetadataReference>()
+        {
+            MetadataReference.CreateFromFile(typeof(Assert).Assembly.Location),
+            MetadataReference.CreateFromFile(typeof(FactAttribute).Assembly.Location),
+            MetadataReference.CreateFromFile(Assembly.Load("System.Runtime, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a").Location), //this feels fragile and wrong but I'm not sure
```
Comment (taylus) [inline Tests/CSharp/SelfVerifyingTests.cs]: Iffy about how these are referenced but couldn't find a better way.

**Comment by taylus** [comment]
Hey @GrahamTheCoder, I think I better understand your latest comments in #210 and pushed some more changes. What do you think? I might be duplicating some compilation code.

**Comment by GrahamTheCoder** [comment]
^ I've rebased it onto master and removed my previous commit - after a git clean, my machine, yours and the build server all agree on the conversion of those attributes.

It's looking really good - I'll make some changes on top and get it merged if possible

**Comment by taylus** [comment]
Np! Thank you for cleaning it up!

**Pull request merged**